### DISCUSSION TITLE
Can't Do `sudo` or `su` From Normal User

### DISCUSSION THREAD

**Discussion by eeriemyxi** [RuriOSS/daijin#54] [Q&A]
I installed an arm64 Arch Linux container with `ruri` on Termux with root. I installed `sudo` and added a user `myxi` then I edited the `/etc/sudoers` file:
```
[root@localhost ~]# cat /etc/sudoers
## sudoers file.
myxi  ALL=(ALL:ALL) ALL
##
## This file MUST be edited with the 'visudo' command as root.
```
I also set a password with `passwd myxi` (I never ran `passwd` standalone)
When I do `su myxi` it logs me in to that user but I can't do `su` there; when I do it asks a password, I tried my user account password but it says authentication failed. I tried not putting anything then hitting enter but same result.

Next, when I do `sudo`, it says this:
```
localhost% sudo
sudo: effective uid is not 0, is /usr/bin/sudo on a file system with the 'nosuid' option set or an NFS file system without root privileges?
```
Here are the file permissions:
```
localhost% ls -l /etc/sudoers
-r--r----- 1 root root 4909 Nov 12 09:14 /etc/sudoers
localhost% ls -l /bin/sudo
-rwsr-xr-x 1 root root 334952 Sep 13 18:45 /bin/sudo
```
I have no idea what's wrong. I want to access what I get when I do `sudo ruri ./arch` when I do `su` from my user account. Or at least make `sudo` work.

**Comment by Moe-hacker** [comment]
You can try running `sudo mount -o remount,suid /data` outside the container.
I added it in https://github.com/Moe-hacker/daijin/commit/c524e76cdd77aeccaf63f21a7767917309c0e6d2

**Reply by eeriemyxi** [reply]
I tried executing that command on my Termux shell after a fresh reboot and doing nothing else before it but I get this:
```
~ $ sudo mount -o remount,suid /data
mount: '/dev/block/dm-0'->'/data': Device or resource busy
~ $
```

**Reply by Moe-hacker** [reply]
You can try  `sudo mount -o remount,suid /path/to/container` after running container, or try `sudo mount --bind -o suid /data /data`.
If you still have the same error, try `unshare -m` , and remount /data & start ruri in the mount namespace.

**Reply by Moe-hacker** [reply]
And as the last choice, If you really need to run sudo but you cannot remount the mountpoint of container, you can migrate your container to an ext4 image, and use `ruri -m image / /path/to/mountpoint` to start.

**Reply by Moe-hacker** [reply]
> I tried executing that command on my Termux shell after a fresh reboot and doing nothing else before it but I get this:
> ```
> ~ $ sudo mount -o remount,suid /data
> mount: '/dev/block/dm-0'->'/data': Device or resource busy
> ~ $
> ```

Btw, can you show me the info of `lsof +D /data` or `fuser -v /data`? I haven't seen bugs like this before.

**Reply by eeriemyxi** [reply]
Hi, I have a feeling that the issues are probably due to incompatibility of my Magisk version. Ever since I re-flashed my stock ROM then flashed Magisk I've noticed some odd things. It never asked me to reboot to finish some extra steps after flashing the patched `boot.img`, then it just crashing whenever I try to install a module that merely symlinks Magisk's BusyBox. I don't know how to downgrade so I've decided to just give up on this whole thing chroot thing. Fuck pi-hole and other stuff I wanted to try I guess.

The version of Magisk I installed was v27, but after flashing my device I installed v28 (I didn't know I had v27 before). So that's probably the reason. I did a `adb logcat` around the time it crashed, https://pastebin.com/uRii2Hgm and it seems like the issue is with the APK not the runtime or whatever Magisk does with the `boot.img`. Sadly I have no idea how to properly and safely downgrade and I have no desire to re-flash my stock ROM for the third time (it takes a long time and I don't have stable electricity connection here).

Now ignoring the venting I just did, could you please report whether this method is safe or not for removing Ruri chroot stuff?

- Reboot the device
- `sudo ruri -U $HOME/ch/arch`
- `pkg remove daijin`
- `cd $HOME/ch`
- `sudo rm -Ir ./arch`
- Reboot the device

Is this procedure safe? Maybe there's a better way?

**Reply by Moe-hacker** [reply]
> Hi, I have a feeling that the issues are probably due to incompatibility of my Magisk version. Ever since I re-flashed my stock ROM then flashed Magisk I've noticed some odd things. It never asked me to reboot to finish some extra steps after flashing the patched `boot.img`, then it just crashing whenever I try to install a module that merely symlinks Magisk's BusyBox. I don't know how to downgrade so I've decided to just give up on this whole thing chroot thing. Fuck pi-hole and other stuff I wanted to try I guess.
> 
> The version of Magisk I installed was v27, but after flashing my device I installed v28 (I didn't know I had v27 before). So that's probably the reason. I did a `adb logcat` around the time it crashed, https://pastebin.com/uRii2Hgm and it seems like the issue is with the APK not the runtime or whatever Magisk does with the `boot.img`. Sadly I have no idea how to properly and safely downgrade and I have no desire to re-flash my stock ROM for the third time (it takes a long time and I don't have stable electricity connection here).
> 
> Now ignoring the venting I just did, could you please report whether this method is safe or not for removing Ruri chroot stuff?
> 
> - Reboot the device
> - `sudo ruri -U $HOME/ch/arch`
> - `pkg remove daijin`
> - `cd $HOME/ch`
> - `sudo rm -Ir ./arch`
> - Reboot the device
> 
> Is this procedure safe? Maybe there's a better way?

It's safe to just reboot & remove, after rebooting, all mountpoints in container will be umounted.

**Reply by Young-Lord** [reply]
> You can try `sudo mount -o remount,suid /path/to/container` after running container, or try `sudo mount --bind -o suid /data /data`. If you still have the same error, try `unshare -m` , and remount /data & start ruri in the mount namespace.

请问可以把这个修复移植到 rurima/ruri 上吗？刚碰到同样的问题翻了一圈才找到答案。

**Reply by Moe-hacker** [reply]
> > You can try `sudo mount -o remount,suid /path/to/container` after running container, or try `sudo mount --bind -o suid /data /data`. If you still have the same error, try `unshare -m` , and remount /data & start ruri in the mount namespace.
> 
> 请问可以把这个修复移植到 rurima/ruri 上吗？刚碰到同样的问题翻了一圈才找到答案。

设置suid位会影响整个/data(父挂载点)，这显然会影响安全性。我不希望为ruri引入任何影响容器目录之外的尤其是安全相关配置的特性。它可以通过外部脚本来解决(如unshare --mount然后mount -o remount,suid /data)，并不需要通过ruri来这样做。
